### ISSUE TITLE
Add a @transform!  mutating version

### ISSUE THREAD

**Issue by jlperla** [JuliaData/DataFramesMeta.jl#110]
In order to port over Stata and R code, it is nice to have succinct calculations of new columns based on existing values.  Right now,
```julia
using DataFrames, DataFramesMeta
df = DataFrame(a=[1; 2; 3], b=[10; 12; 13])
f(a, b) = a + b
df =@transform(df, c = f.(:a, :b))
```
But it would be great if this could instead be:
```julia
using DataFrames, DataFramesMeta
df = DataFrame(a=[1; 2; 3], b=[10; 12; 13])
f(a, b) = a + b
@transform!(df, c = f.(:a, :b))
```

Is this a relatively easy change for someone to add a PR?

**Comment by jlperla** [comment]
Thanks for the idea.  Actually, let me make it even better, since variadic makes a lot of sense here.
```julia
using DataFrames, DataFramesMeta
df = DataFrame(a=[1; 2; 3], b=[10; 12; 13])
f(a, b) = a + b
@transform!(df, c = f.(:a, :b), d = :a + :b)
```
instead of
```julia
using DataFrames, DataFramesMeta
df = DataFrame(a=[1; 2; 3], b=[10; 12; 13])
f(a, b) = a + b
df =@transform(df, c = f.(:a, :b))
df =@transform(df, d = :a + :b)
```

**Comment by pdeffebach** [comment]
One thing to note is that if you are doing a mutating transforms over and over again, `@with` is probably a more concise macro to use. 

```
@with df begin 
    df.t = f.(:a, :b)
    df.t = :a + :b
end
```

**Comment by pdeffebach** [comment]
This would actually a bit tough for a grouped data frame. Because `transform(g::GroupedDataFrame, args...)` uses `DataFrame(g)` in the first line, it's always going to make a copy. Perhaps we could do 

```
d = sort!(g.parent, g.cols)
..all the other stuff
```

**Comment by pdeffebach** [comment]
I just created a PR for this, but I think we should have more discussion about whether or not it is the right strategy. 

It's nice that this is an opinionated package where there is one way of doing each thing. I might prefer to have just one, albeit copying, macro for each operation than face a splintered API. 

Despite opening the PR I might wonder if we can convince users to get used to copying, and make sure that everything is as performant as possible.

**Comment by matthieugomez** [comment]
Having in-place versions of operations that require a deep copy of the dataframe would be particularly useful: it would make it possible to work with dataframes that are close to the maximum amount of RAM.  If I understand correctly, these would be `@orderby` and `@where`.

**Comment by pdeffebach** [comment]
Yes I agree. Though I haven't heard too many complaints to this effect with `dplyr`. 

I just added a commit in #119 making some more progress. `@orderby` and `@where` are on the list

**Comment by pdeffebach** [comment]
Closed via #216
